### PR TITLE
#15950 Fix the behavior of toggling line comment 

### LINKED ISSUE #15950: Toogle line comment changes selection
#### System information: 
- Operating system (distribution) and version: Windows 10 21H1
- DBeaver version: 22.0.1
- Additional extensions: No

#### Connection specification:
- Database name and version: Snowflake, MS SQL Server, Oracle
- Driver name: Not relevant
- Yes

#### Describe the problem you're observing:
The problem happens in the SQL editor. It is independend of the database connection.
If you mark some lines and press Ctrl+B (Toogle line comment), the lines will be commented and the position of the begin selection changes (but not correct).
If you press Ctrl+B again, the lines will be uncommented, and the selection will move to a complete different possition (for begin and end).
In earlier versions, it worked fine!

#### Steps to reproduce, if exist:
Mark some lines
Ctrl+B
Ctrl+B

### PR DESCRIPTION
**Pull request by arvillion**
Fix issue #15950 

No idea how to write and where to put test cases. So I write some possible test scenarios.
Note that in case 1, the comment string(`--`) on the first line is not focused, which is the same as what VSCode will do.

1. Comment, and then the first comment string should not be selected
2. Uncomment with the first line fully selected
3. Uncomment with the first line partially selected and exclude the first comment string
4. Uncomment with the first line partially selected and include the first comment string
5. Uncomment with some intervening blank lines

https://user-images.githubusercontent.com/3462292/163926717-b2c72569-c4b6-4d1a-89e4-bf658fab5ec1.mp4

https://user-images.githubusercontent.com/3462292/163926721-866021a6-47e0-4832-a8c9-68945f41932a.mp4

https://user-images.githubusercontent.com/3462292/163926722-211740d5-418e-40e2-9864-f1b3692ee0ae.mp4

https://user-images.githubusercontent.com/3462292/163926724-a7552f4e-6389-4cfe-b8e4-c2e8800bc2bf.mp4

https://user-images.githubusercontent.com/3462292/163926726-cdb59bfd-566a-42d0-a1ea-1007d509795b.mp4

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/text/handlers/ToggleSingleLineCommentHandler.java [hunk 1]
```diff
@@ -48,13 +50,18 @@ protected void processAction(ISelectionProvider selectionProvider, ICommentsSupp
...
             int lineLength = document.getLineLength(lineNum);
             if (forceComment) {
                 // Add comment
                 document.replace(lineOffset, 0, singleLineComments[0]);
-                selOffset += singleLineComments[0].length();
+                if(lineNum == startLine) {
+                	selOffset += singleLineComments[0].length();
+                } else {
+                	selLength += singleLineComments[0].length();
+                }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/text/handlers/ToggleSingleLineCommentHandler.java]: Please fix the formatting. We don't use tabs for indentation, and there's a missing space after the `if` keyword.

File: plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/text/handlers/ToggleSingleLineCommentHandler.java [hunk 2]
```diff
@@ -75,7 +82,16 @@ protected void processAction(ISelectionProvider selectionProvider, ICommentsSupp
                 if (lineComment != null) {
                     // Remove comment
                     document.replace(lineOffset + checkOffset, lineComment.length(), "");
-                    selOffset = Math.max(lineOffset, selOffset - lineComment.length());
+                    
+                    if (lineNum == startLine && originalOffset != lineOffset) {
+                    	selOffset = Math.max(lineOffset, originalOffset - lineComment.length());
+                    	if(lineComment.length() + lineOffset > originalOffset) {
+                    		selLength -= lineComment.length() - originalOffset + lineOffset;
+                    	}
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/text/handlers/ToggleSingleLineCommentHandler.java]: Same as the previous comment.

File: plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/text/handlers/ToggleSingleLineCommentHandler.java [hunk 3]
```diff
@@ -84,6 +100,7 @@ protected void processAction(ISelectionProvider selectionProvider, ICommentsSupp
         }
         if (originalLength > 0) {
             selectionProvider.setSelection(new TextSelection(selOffset, selLength));
+
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/text/handlers/ToggleSingleLineCommentHandler.java]: This extra line should not be there.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Have you verified that the block selection mode works just fine as a regular selection?

You can enable it using the <kbd>Ctrl+Shift+A</kbd> shortcut or find it under the <kbd>Edit</kbd> menu.

**Comment by arvillion** [comment]
> Thanks for the contribution!
> 
> Have you verified that the block selection mode works just fine as a regular selection?
> 
> You can enable it using the Ctrl+Shift+A shortcut or find it under the Edit menu.

The block selection works correctly.

https://user-images.githubusercontent.com/3462292/164038339-0e0bc455-0096-4cb5-86a3-ba0757773c2a.mp4

**Review by ShadelessFox — APPROVED** [review]
👍🏻

**Comment by bianyun1981** [comment]
@arvillion Great work!  Thanks for your contribution.

**Comment by ShadelessFox** [comment]
The pull request was removed from the milestone because its state is tracked via the linked issue.

**Comment by DariaMarkaryan** [comment]
The strange behavior of the selection is still observed. Perhaps the same as you had before, I can’t watch the video attached to the PR

https://user-images.githubusercontent.com/43252954/164750911-4ec8d77a-9737-4e6f-87f1-a5f58ceb59d3.mp4

**Comment by arvillion** [comment]
@DariaMarkaryan 

Well, I tried out the same selection with the same text as yours. And it just works fine.

**Comment by arvillion** [comment]
@DariaMarkaryan 

Well, I tried out the same selection with the same text as yours and it just works fine. Do you mean that this problem still exist in current devel branch?

**Comment by Matvey16** [comment]
Multiple presses of Ctrl+/ will move the selecteon until it's at the end

dbeaver_dNPklYNdQ3

**Comment by arvillion** [comment]
@Matvey16
8589z-cdt7s

**Comment by ShadelessFox** [comment]
Hello @arvillion,

I have verified that your fix is working just fine.

I believe there are some problems with our CI, especially with external PRs. We'll look into that.

**Comment by Matvey16** [comment]
Verified

**Pull request merged**